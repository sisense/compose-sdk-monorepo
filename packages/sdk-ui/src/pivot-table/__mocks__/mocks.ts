import * as DM from '../../__demo__/sample-ecommerce';
import { filterFactory, measureFactory } from '@sisense/sdk-data';
import { PivotTableProps } from '../../props';

export const mockPivotTableProps: PivotTableProps = {
  dataSet: DM.DataSource,
  dataOptions: {
    rows: [
      { column: DM.Category.Category, includeSubTotals: true },
      { column: DM.Commerce.AgeRange, includeSubTotals: true },
      DM.Commerce.Condition,
    ],
    columns: [{ column: DM.Commerce.Gender, includeSubTotals: true }],
    values: [
      {
        column: measureFactory.sum(DM.Commerce.Cost, 'Total Cost'),
        dataBars: true,
        totalsCalculation: 'sum',
      },
      {
        column: measureFactory.sum(DM.Commerce.Revenue, 'Total Revenue'),
        totalsCalculation: 'sum',
      },
    ],
    grandTotals: { title: 'Grand Total', rows: true, columns: true },
  },
  filters: [filterFactory.members(DM.Commerce.Gender, ['Female', 'Male'])],
};
