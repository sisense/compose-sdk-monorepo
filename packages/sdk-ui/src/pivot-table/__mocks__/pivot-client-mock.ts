/* eslint-disable no-unused-vars */
/* eslint-disable sonarjs/no-identical-functions */
import React from 'react';
import { PivotClient } from '@ethings-os/sdk-pivot-client';

// Mock of PivotBuilder class
class MockPivotBuilder {
  private listeners: Record<string, Array<(...args: any[]) => void>> = {};

  on(eventName: string, callback: (...args: any[]) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(callback);
  }

  off(eventName: string, callback: (...args: any[]) => void) {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((cb) => cb !== callback);
    }
  }

  emit(eventName: string, ...args: any[]) {
    if (this.listeners[eventName]) {
      this.listeners[eventName].forEach((callback) => callback(...args));
    }
  }

  render(props: any) {
    // Mock render method - does not actually render anything
    // Simulate emitting a proper React element instead of a string
    const mockElement = React.createElement(
      'div',
      { 'data-testid': 'mocked-pivot-table' },
      'Mocked Pivot Table',
    );
    setTimeout(() => {
      this.emit('pivotElementChange', mockElement);
    }, 0);
  }

  updateJaql(jaql: any) {
    // Mock updateJaql method
    setTimeout(() => {
      this.emit('queryStart');
      // Simulate query end with empty metadata
      setTimeout(() => {
        this.emit('queryEnd', { cellsMetadata: null });
      }, 10);
    }, 0);
  }

  updateDataService(dataService: any) {
    // Mock updateDataService method
  }

  destroy() {
    // Mock destroy method
    this.listeners = {};
  }
}

// Mock of DataService class
class MockDataService {
  private listeners: Record<string, Array<(...args: any[]) => void>> = {};

  on(eventName: string, callback: (...args: any[]) => void) {
    if (!this.listeners[eventName]) {
      this.listeners[eventName] = [];
    }
    this.listeners[eventName].push(callback);
  }

  off(eventName: string, callback: (...args: any[]) => void) {
    if (this.listeners[eventName]) {
      this.listeners[eventName] = this.listeners[eventName].filter((cb) => cb !== callback);
    }
  }

  destroy() {
    // Mock destroy method
    this.listeners = {};
  }
}

// Create a mocked PivotClient
export const createMockPivotClient = () => {
  return {
    preparePivotBuilder: vi.fn(() => new MockPivotBuilder()),
    prepareDataService: vi.fn(() => new MockDataService()),
    queryData: vi.fn(),
    socketBuilder: {
      socket: {},
    },
  } as unknown as PivotClient;
};
