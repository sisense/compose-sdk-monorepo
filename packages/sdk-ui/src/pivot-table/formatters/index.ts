export * from './data-cell-formatters';
export * from './header-cell-formatters';
