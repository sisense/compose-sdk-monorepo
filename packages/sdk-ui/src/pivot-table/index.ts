export { PivotTable } from './pivot-table';
