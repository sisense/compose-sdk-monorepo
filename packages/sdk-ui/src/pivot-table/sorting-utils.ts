import { JaqlSortDirection, getSortType } from '@sisense/sdk-data';
import { type SortingSettingsChangePayload } from '@sisense/sdk-pivot-client';
import isEqual from 'lodash-es/isEqual';
import isString from 'lodash-es/isString';
import { PivotTableDataOptionsInternal } from '..';
import type { SortDirection, PivotRowsSort } from '../types';

export type PivotRowsSortCriteria = {
  rowIndex: number;
  sort: PivotRowsSort | null;
};

export type PivotSortCriteria = {
  rows: PivotRowsSortCriteria[];
};

export function normalizePivotSort(sortType: SortDirection | PivotRowsSort) {
  return isString(sortType) ? { direction: sortType } : sortType;
}

function getSortBy(
  sortDetails: SortingSettingsChangePayload['sortDetails'],
  dataOptionsInternal: PivotTableDataOptionsInternal,
) {
  return {
    valuesIndex:
      sortDetails.field! -
      (dataOptionsInternal.rows || []).length -
      (dataOptionsInternal.columns || []).length,
    columnsMembersPath: Object.values(sortDetails.measurePath || {}),
  };
}

export function preparePivotRowsSortCriteriaList(
  { type, settings, sortDetails }: SortingSettingsChangePayload,
  dataOptionsInternal: PivotTableDataOptionsInternal,
): PivotRowsSortCriteria[] {
  if (type === 'simple') {
    const isSortedByMeasure = sortDetails.field! >= (dataOptionsInternal.rows || []).length;

    if (isSortedByMeasure) {
      return [
        {
          rowIndex: 0,
          sort: {
            direction: getSortType(settings[0].direction as JaqlSortDirection),
            by: getSortBy(sortDetails, dataOptionsInternal),
          },
        },
      ];
    }

    return [
      {
        rowIndex: sortDetails.field!,
        sort: {
          direction: getSortType(settings[0].direction as JaqlSortDirection),
        },
      },
    ];
  }

  // 'complex' sorting type
  return settings.reduce((sortCriteriaList: PivotRowsSortCriteria[], sortOptions, index) => {
    const rowDataOption = dataOptionsInternal.rows?.[index];
    const existingRowSort = rowDataOption?.sortType && normalizePivotSort(rowDataOption?.sortType);
    const newSort = {
      direction: getSortType(sortOptions.direction as JaqlSortDirection),
      by: getSortBy(sortDetails, dataOptionsInternal),
    };

    const shouldDisableRowSort = !sortOptions.selected && isEqual(existingRowSort?.by, newSort.by);

    if (shouldDisableRowSort) {
      sortCriteriaList.push({
        rowIndex: index,
        sort: null,
      });
    }

    if (sortOptions.selected) {
      sortCriteriaList.push({
        rowIndex: index,
        sort: newSort,
      });
    }

    return sortCriteriaList;
  }, []);
}
