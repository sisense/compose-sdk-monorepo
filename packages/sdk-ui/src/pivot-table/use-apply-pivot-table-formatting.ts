/* eslint-disable react-hooks/exhaustive-deps */
import { useCallback, useEffect } from 'react';
import {
  type DataService,
  EVENT_DATA_CELL_FORMAT,
  EVENT_HEADER_CELL_FORMAT,
} from '@sisense/sdk-pivot-client';
import over from 'lodash/over';
import { type PivotTableDataOptions } from '@/chart-data-options/types';
import { applyDateFormat } from '@/query/date-formats';
import { useSisenseContext } from '@/sisense-context/sisense-context';
import { createDataCellValueFormatter, createHeaderCellValueFormatter } from './formatters';
import { createHeaderCellHighlightFormatter } from './formatters/header-cell-formatters/header-cell-highlight-formatter';

/**
 * A hook that applies formatting over pivot table cells.
 *
 * @internal
 */
export const useApplyPivotTableFormatting = ({
  dataService,
  dataOptions,
}: {
  dataService: DataService;
  dataOptions: PivotTableDataOptions;
}) => {
  const { app } = useSisenseContext();

  const onDataCellFormat = useCallback(over([createDataCellValueFormatter(dataOptions)]), [
    dataOptions,
  ]);

  const dateFormatter = useCallback(
    (date: Date, format: string) =>
      applyDateFormat(date, format, app?.settings.locale, app?.settings.dateConfig),
    [app],
  );

  const onHeaderCellFormat = useCallback(
    over([
      createHeaderCellValueFormatter(dataOptions, dateFormatter),
      createHeaderCellHighlightFormatter(),
    ]),
    [dataOptions],
  );

  useEffect(() => {
    dataService.on(EVENT_DATA_CELL_FORMAT, onDataCellFormat);
    dataService.on(EVENT_HEADER_CELL_FORMAT, onHeaderCellFormat);
    return () => {
      dataService.off(EVENT_DATA_CELL_FORMAT, onDataCellFormat);
      dataService.off(EVENT_HEADER_CELL_FORMAT, onHeaderCellFormat);
    };
  }, [dataService, onDataCellFormat]);
};
