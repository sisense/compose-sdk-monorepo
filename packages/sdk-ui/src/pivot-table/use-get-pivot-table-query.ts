import { useCallback, useRef } from 'react';
import { type JaqlRequest } from '@sisense/sdk-pivot-client';
import type { DataSource, Filter, FilterRelations } from '@sisense/sdk-data';
import { ExecutePivotQueryParams } from '../query-execution';
import { useExecutePivotQueryInternal } from '../query-execution/use-execute-pivot-query';
import { Category, PivotTableDataOptionsInternal, Value } from '../chart-data-options/types';
import { translateCategoryToAttribute, translateValueToMeasure } from '../chart-data-options/utils';
import { normalizePivotSort } from './sorting-utils';

const getPivotAttribute = (category: Category) => {
  return {
    attribute: translateCategoryToAttribute(category),
    includeSubTotals: category.includeSubTotals || false,
    ...(category.sortType && {
      sort: normalizePivotSort(category.sortType),
    }),
  };
};

const getPivotMeasure = (value: Value) => {
  return {
    measure: translateValueToMeasure(value),
    totalsCalculation: value.totalsCalculation,
    dataBars: value.dataBars || false,
  };
};

export const getPivotQueryOptions = (
  dataOptions: PivotTableDataOptionsInternal,
): Pick<ExecutePivotQueryParams, 'rows' | 'columns' | 'values' | 'grandTotals'> => {
  const {
    rows: rowsInternal = [],
    columns: columnsInternal = [],
    values: valuesInternal = [],
    grandTotals,
  } = dataOptions;

  const rows = rowsInternal.flatMap(getPivotAttribute);
  const columns = columnsInternal.flatMap(getPivotAttribute);
  const values = valuesInternal.flatMap(getPivotMeasure);

  return { rows, columns, values, grandTotals };
};

/**
 * A hook that returns the JAQL request object from pivot table props.
 *
 * @internal
 */
export const useGetPivotTableQuery = ({
  dataSet,
  dataOptionsInternal,
  filters,
  refreshCounter,
}: {
  dataSet?: DataSource;
  dataOptionsInternal: PivotTableDataOptionsInternal;
  filters?: Filter[] | FilterRelations;
  refreshCounter?: number;
}) => {
  const jaqlRef = useRef<JaqlRequest | null>(null);
  const { rows, columns, values, grandTotals } = getPivotQueryOptions(dataOptionsInternal);

  // retrieve the Jaql query without executing it
  const onBeforeQuery = useCallback(
    (query: JaqlRequest) => {
      jaqlRef.current = query;
    },
    // forces 'onBeforeQuery' callback to be recreated on 'refreshCounter' change in order to refresh the jaql
    // eslint-disable-next-line react-hooks/exhaustive-deps
    [refreshCounter],
  );

  const queryParams: ExecutePivotQueryParams = {
    dataSource: dataSet,
    rows,
    columns,
    values,
    grandTotals,
    filters,
    onBeforeQuery,
  };
  const { isLoading, isSuccess, isError, error } = useExecutePivotQueryInternal(queryParams);

  return { isLoading, isSuccess, isError, error, jaql: jaqlRef.current };
};
