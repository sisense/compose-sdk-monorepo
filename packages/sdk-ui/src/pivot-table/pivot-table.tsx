import React, { useEffect, useRef } from 'react';
import { PivotTableProps } from '../props';
import { asSisenseComponent } from '../decorators/component-decorators/as-sisense-component';
import { shouldSkipSisenseContextWaiting } from '../chart';
import { useSisenseContext } from '../sisense-context/sisense-context';
import { useGetPivotTableQuery } from './use-get-pivot-table-query';

const DEFAULT_TABLE_ROWS_PER_PAGE = 25 as const;

/**
 * Pivot Table with pagination.
 *
 * See [Pivot Tables](https://docs.sisense.com/main/SisenseLinux/pivot.htm) for more information.
 *
 * @example
 * (1) Example of PivotTable from the `Sample ECommerce` data model:
 *
 * ```tsx
 * <PivotTable
 *   dataSet={DM.DataSource}
 *   dataOptions={{
 *     rows: [
 *       { column: DM.Category.Category, includeSubTotals: true },
 *       { column: DM.Commerce.AgeRange, includeSubTotals: true },
 *       DM.Commerce.Condition,
 *     ],
 *     columns: [{ column: DM.Commerce.Gender, includeSubTotals: true }],
 *     values: [
 *       {
 *         column: measureFactory.sum(DM.Commerce.Cost, 'Total Cost'),
 *         dataBars: true,
 *         totalsCalculation: 'sum',
 *       },
 *       {
 *         column: measureFactory.sum(DM.Commerce.Revenue, 'Total Revenue'),
 *         totalsCalculation: 'sum',
 *       },
 *     ],
 *     grandTotals: { title: 'Grand Total', rows: true, columns: true },
 *   }}
 *   filters={[filterFactory.members(DM.Commerce.Gender, ['Female', 'Male'])]}
 *   styleOptions={{ width: 1000, height: 600, rowsPerPage: 50 }}
 * />
 * ```
 * @param props - Pivot Table properties
 * @returns Pivot Table component
 * @alpha
 */

export const PivotTable = asSisenseComponent({
  componentName: 'PivotTable',
  shouldSkipSisenseContextWaiting,
})((pivotTableProps: PivotTableProps) => {
  // retrieve and validate the pivot client
  const { app } = useSisenseContext();
  const pivotClient = app?.pivotClient;
  if (!pivotClient) {
    throw new Error('Pivot client not initialized');
  }

  const { styleOptions = {} } = pivotTableProps;
  // get the jaql from the pivot table props
  const { isError, error, jaql } = useGetPivotTableQuery(pivotTableProps);

  if (isError) {
    throw error;
  }

  const nodeRef = useRef<HTMLDivElement>(null);
  const pivotBuilder = pivotClient.preparePivotBuilder();

  useEffect(() => {
    if (nodeRef.current && jaql) {
      const {
        rowsPerPage = DEFAULT_TABLE_ROWS_PER_PAGE,
        width = 400,
        height = 400,
        isAutoHeight = false,
      } = styleOptions;
      const props = {
        width,
        height,
        isAutoHeight,
        isPaginated: true,
        itemsPerPage: rowsPerPage,
      };

      pivotBuilder.render(nodeRef.current, props);
      pivotBuilder.updateJaql(jaql);
    }

    return () => {
      // for cleanup if needed
    };
  }, [jaql, pivotBuilder, pivotClient, styleOptions]);

  return <div ref={nodeRef} aria-label="pivot-table-root" />;
});
