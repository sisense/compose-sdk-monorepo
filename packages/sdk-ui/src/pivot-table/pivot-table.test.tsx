/* eslint-disable @typescript-eslint/no-unsafe-call */
/* eslint-disable @typescript-eslint/no-unsafe-member-access */
/* eslint-disable @typescript-eslint/no-unsafe-assignment */
/** @vitest-environment jsdom */
import { fireEvent, render } from '@testing-library/react';
import { PivotTable } from './pivot-table';
import { mockPivotTableProps } from './__mocks__/mocks';
import { useSisenseContextMock } from '../sisense-context/__mocks__/sisense-context';
import { ClientApplication } from '../app/client-application';
import { PivotClient } from '@sisense/sdk-pivot-client';
import { SsoAuthenticator } from '@sisense/sdk-rest-client';
import { executePivotQueryMock } from '../query/__mocks__/execute-query';
import { EMPTY_PIVOT_QUERY_RESULT_DATA } from '@sisense/sdk-data';
import { SisenseContextPayload } from '@/sisense-context/sisense-context';

vi.mock('../query/execute-query');
vi.mock('../sisense-context/sisense-context');

describe('PivotTable', () => {
  beforeEach(() => {
    const url = 'mock-url';
    const contextMock: SisenseContextPayload = {
      app: {
        pivotClient: new PivotClient(url, new SsoAuthenticator(url), true),
        settings: {
          trackingConfig: { enabled: false },
        },
      } as ClientApplication,
      isInitialized: true,
      tracking: {
        enabled: false,
        packageName: '',
      },
      errorBoundary: {
        showErrorBox: true,
      },
    };
    useSisenseContextMock.mockReturnValue(contextMock);
  });
  it('should render empty pivot table', async () => {
    executePivotQueryMock.mockResolvedValue(EMPTY_PIVOT_QUERY_RESULT_DATA);
    const { container, findByLabelText } = render(<PivotTable {...mockPivotTableProps} />);
    const pivotTable = await findByLabelText('pivot-table-root');
    expect(pivotTable).toBeTruthy();
    expect(container).toMatchSnapshot();
  });

  it('should handle error', async () => {
    const spy = vi.spyOn(console, 'error');
    spy.mockImplementation(() => {});

    const mockError = new Error('Test error');
    executePivotQueryMock.mockRejectedValue(mockError);
    const { findByLabelText, findByText } = render(<PivotTable {...mockPivotTableProps} />);

    const errorBoxContainer = await findByLabelText('error-box');
    fireEvent.mouseEnter(errorBoxContainer);
    const errorBoxText = await findByText(/errorBoxText/);
    expect(errorBoxText).toBeTruthy();

    spy.mockRestore();
  });
});
