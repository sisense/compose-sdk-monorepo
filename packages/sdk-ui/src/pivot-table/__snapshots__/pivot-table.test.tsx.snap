// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PivotTable > should render empty pivot table 1`] = `
<div>
  <div
    aria-label="pivot-table-root"
  />
</div>
`;
